src/main/host/descriptor/file.c

```diff
@@ -194,87 +183,118 @@ static char* _file_getPath(File* file, File* dir, const char* pathname) {
-         * an absolute path to compare for special files. */
-        abspath = _file_getPath(file, dir, pathname);
-
-        /* TODO: Handle special files. */
-        if (utility_isRandomPath(abspath)) {
-            file->type = FILE_TYPE_RANDOM;
-        } else if (!strcmp("/etc/hosts", abspath)) {
-            file->type = FILE_TYPE_HOSTS;
-            char* hostspath = dns_getHostsFilePath(worker_getDNS());
-            if(hostspath && abspath) {
-                free(abspath);
-                abspath = hostspath;
-            }
-        } else if (!strcmp("/etc/localtime", abspath)) {
-            file->type = FILE_TYPE_LOCALTIME;
-        } else {
-            file->type = FILE_TYPE_REGULAR;
-        }
+        flags = (flags & ~O_EXCL);
+    }
 
-        if (file->type == FILE_TYPE_LOCALTIME) {
-            // Fail the localtime lookup so the plugin falls back to UTC.
-            // TODO: we could instead return a special file that contains
-            // timezone info in the correct format for UTC.
-            osfd = -1;
-            errno = ENOENT;
-        } else {
-            osfd = open(abspath, flags, mode);
+    /* Handle special files. */
+    if (utility_isRandomPath(abspath)) {
+        file->type = FILE_TYPE_RANDOM;
+    } else if (!strcmp("/etc/hosts", abspath)) {
+        file->type = FILE_TYPE_HOSTS;
+        char* hostspath = dns_getHostsFilePath(worker_getDNS());
+        if (hostspath && abspath) {
+            free(abspath);
+            abspath = hostspath;
         }
+    } else if (!strcmp("/etc/localtime", abspath)) {
+        file->type = FILE_TYPE_LOCALTIME;
+    } else {
+        file->type = FILE_TYPE_REGULAR;
+    }
+
+    int osfd = 0, errcode = 0;
+    if (file->type == FILE_TYPE_LOCALTIME) {
+        // Fail the localtime lookup so the plugin falls back to UTC.
+        // TODO: we could instead return a special file that contains
+        // timezone info in the correct format for UTC.
+        osfd = -1;
+        errcode = ENOENT;
+    } else {
+        // TODO: we should open the os-backed file in non-blocking mode even if a
+        // non-block is not requested, and then properly handle the io by, e.g.,
+        // epolling on all such files with a shadow support thread.
+        osfd = open(abspath, flags, mode);
+        errcode = errno;
     }
 
     if (osfd < 0) {
-        debug("File %i opening path '%s' returned %i: %s", _file_getFD(file),
-              abspath, osfd, strerror(errno));
+        debug("File %i opening path '%s' returned %i: %s", _file_getFD(file), abspath, osfd,
+              strerror(errcode));
         if (abspath) {
             free(abspath);
         }
         file->type = FILE_TYPE_NOTSET;
-        return -errno;
+        return -errcode;
     }
 
-    /* Store the create information so we can open later if needed. */
+    /* Store the create information so we can re-open later if needed (e.g., mmap). */
```

Comment:
Still needed?

Reply:
We use `osfd`, `flags`, and `mode` when we `mmap` the file, and `abspath` for logging.